File: CourseProjects/CourseProjects/VisualStudioProjects/Languages/Basic VB Programming/Inheritance/Employee.vb
Imports System.Text

Public Class Employee
    'Class constructors
    Public Sub New(ByVal id As String, ByVal name As String, ByVal age As Integer, ByVal salary As Double)
        Me.New(id, name, age)
        Me.salary = salary
    End Sub
    Public Sub New(ByVal id As String, ByVal name As String, ByVal age As Integer)
        Me.New(id, name)
        Me.age = age
    End Sub
    Public Sub New(ByVal id As String, ByVal name As String)
        Me.id = id
        Me.name = name
    End Sub

    'Show polymorphism
    Public Overridable Sub awardBonus(ByVal bonus As Double)
        salary = salary + bonus
    End Sub

    'Show overloading
    Public Overloads Sub markOnHoliday()
        onHoliday = True
    End Sub

    Public Overloads Sub markOnHoliday(ByVal days As Integer)
        markOnHoliday()
        holidaysTaken += days
    End Sub

    'Override ToString from Object
    Public Overrides Function ToString() As String
        Dim sb As StringBuilder = New StringBuilder()
        sb.Append(id)
        sb.Append(" called ")
        sb.Append(name)
        sb.Append(" aged ")
        sb.Append(age)
        sb.Append(" earning ")
        sb.Append(salary)

        ToString = sb.ToString()
    End Function

    'Fields of the class
    Protected name, id As String
    Protected age As Integer
    Protected salary As Double
    Protected onHoliday As Boolean
    Protected holidaysTaken As Integer


End Class

File: CourseProjects/CourseProjects/VisualStudioProjects/Languages/Basic CS  Programming/Inheritance/Employee.cs
using System;

namespace InheritanceDemo
{
	public class Employee {

		public Employee(string id, string name) : base() {
			this.id = id;
			this.name = name;
		}

		public Employee(string id, string name, int age) : this(id,name) {
			this.age = age;
		}


		public Employee(string id, string name, int age, double salary) : this(id,name,age) {
			this.salary = salary;
		}

		public void markOnHoliday(int days) {
			markOnHoliday();
			holidaysTaken += days;
		}

		public void markOnHoliday() {
			onHoliday = true;
		}

		//method must be virtual to switch on polymorphism
		public virtual void awardBonus(double bonus) {
			salary += bonus;
		}

		public override string ToString() {
			return id + " called " + name + " aged " + age + " earning " + salary;
		}

		//fields
		protected string id;
		protected string name;
		protected int age;
		protected double salary;
		protected bool onHoliday;
		protected int holidaysTaken;
	}
}
